### PR TITLE
Shard MachinePool controller

### PR DESCRIPTION
**Pull request by 2uasimojo**
- Split the machinepool controller out of the main hive-controllers.
- Deploy it via a StatefulSet.
- Make each replica handle a deterministic subset of MachinePools, assigned by matching the ordinal ID of the StatefulSet replica to the modulo of the MachinePool's UID by the number of replicas.
- Record the assigned replica's ordinal ID in a new `MachinePool.Status.ControlledByReplica` field.

This will allow the hive consumer to scale the MachinePool controller via `HiveConfig.Spec.ControllersConfig.Controllers[?.Name=="machinepool"].Config`:
- Horizontally via `.Replicas`.
- Vertically via the other sub-fields, including `.Resources`.

[HIVE-2537](https://issues.redhat.com//browse/HIVE-2537)

### CODE REVIEW

**Comment by 2uasimojo** [comment]
/hold

UT pending. I'm putting this up to see if I can get a proper built image I can use for testing.

**Comment by 2uasimojo** [comment]
/override ci/prow/security

#2345 

/override ci/prow/periodic-images

https://github.com/openshift/release/pull/54083

/override "Red Hat Konflux / hive-on-pull-request"

:shrug: I think @celebdor knows what's happening here?

**Comment by 2uasimojo** [comment]
> /hold
> 
> UT pending. I'm putting this up to see if I can get a proper built image I can use for testing.

All good here. Going to keep the hold for now so we can control merge on (hopefully) Monday.

File: pkg/controller/clustersync/clustersync_controller_test.go [hunk 1]
```diff
@@ -576,15 +394,10 @@ func TestReconcileClusterSync_ApplySecret(t *testing.T) {
...
-			rt := newReconcileTest(t, mockCtrl, scheme,
-				cdBuilder(scheme).Build(),
-				clusterSyncBuilder(scheme).Build(),
-				teststatefulset.FullBuilder("hive", stsName, scheme).Build(
-					teststatefulset.WithCurrentReplicas(3),
-					teststatefulset.WithReplicas(3),
-				),
-				syncSet,
-				srcSecret)
+			rt := func() *reconcileTest {
```
Comment (2uasimojo) [inline pkg/controller/clustersync/clustersync_controller_test.go:397]: ntr: Well, I was just trying to de-lint this thing to remove unused vars (t, scheme). I used a vscode widget to do the refactor... and it did this (throughout). It's not _wrong_, but it's kinda goofy :(

File: pkg/operator/hive/sharded_controllers.go [hunk 2]
```diff
@@ -74,89 +130,85 @@ func (r *ReconcileHiveConfig) deployClusterSync(hLog log.FieldLogger, h resource
...
 	}
 
 	hiveNSName := GetHiveNamespace(hiveconfig)
 
 	// Load namespaced assets, decode them, set to our target namespace, and apply:
-	for _, assetPath := range namespacedAssets {
-		if err := util.ApplyAssetWithNSOverrideAndGC(h, assetPath, hiveNSName, hiveconfig); err != nil {
-			hLog.WithError(err).Error("error applying object with namespace override")
+	for _, a := range namespacedAssets {
+		// YOU ARE ALSO HERE: convert this func to take bytes, like you did for delete
```
Comment (suhanime) [inline pkg/operator/hive/sharded_controllers.go:144]: Intentional?

File: pkg/controller/utils/statefulset.go [hunk 3]
```diff
@@ -42,3 +55,87 @@ func CalculateStatefulSetSpecHash(statefulset *appsv1.StatefulSet) (string, erro
...
+		return false, err
+	}
+
+	if sts.Spec.Replicas == nil {
+		return false, errors.New("sts.Spec.Replicas not set")
+	}
+
+	if sts.Status.CurrentReplicas != *sts.Spec.Replicas {
+		// This ensures that we don't have partial syncing which may make it seem like things are working.
+		// TODO: Remove this once https://issues.redhat.com/browse/CO-1268 is completed as this should no longer be needed.
```
Comment (suhanime) [inline pkg/controller/utils/statefulset.go:108]: It still redirects fine, but Cluster Operator project has since been renamed to Hive.

**Review by suhanime — COMMENTED** [review]
Minor changes, but to my limited understanding, looks good.

**Comment by suhanime** [comment]
/lgtm

Minor corrections to be addressed in a follow-up

File: pkg/controller/utils/statefulset.go (hunk 3 above)
Comment (2uasimojo) [reply pkg/controller/utils/statefulset.go:108]: Yah; this comment was just ported over. I left it untouched for, um, "posterity" :P

File: pkg/operator/hive/sharded_controllers.go (hunk 2 above)
Comment (2uasimojo) [reply pkg/operator/hive/sharded_controllers.go:144]: Heh, thanks, a note-to-self as I left my desk for the night, forgot to scrub it. Will do in a fup.

File: pkg/test/machinepool/machinepool.go [hunk 4]
```diff
@@ -121,6 +122,14 @@ func WithInitializedStatusConditions() Option {
 	}
 }
 
+// WithInitializedStatusConditions returns an Option that *replaces* status conditions
```
[the remark this answers is not in the document]
Comment (2uasimojo) [reply pkg/test/machinepool/machinepool.go:125]: copypasta, will fix, good eye.

File: pkg/controller/clustersync/clustersync_controller_test.go (hunk 1 above)
Comment (2uasimojo) [reply pkg/controller/clustersync/clustersync_controller_test.go:397]: Looking at it again today, I'm actually liking the refactored version. Certainly not worth the churn to convert it back. Going to leave it alone.

**Pull request merged**